### DISCUSSION TITLE
Highlighted selection - not to lose focus when focusing on other element on the page

### DISCUSSION THREAD

**Discussion by demiro** [ueberdosis/tiptap#4963] [General Questions]
I know it sounds a bit confusing...

but inherently the selection get lost when you focus on another element

imagine this scenario:
user selects some text
clicks on a chatbot window and ask the bot something related to the selection

obviously as soon as you click into checkbox window, the editor's selection gets lost.... but, is it possible to wrap the whole selection in a special <span class="highlighted-text">..., and when I select something else, it gets unwrapped?

I saw some other editors do it, not sure how would I go about it in tiptap? Is there an extension or command that would do that?

**Comment by abecirovic3** [comment]
I'm interested in this as well. I have a similar use case, how I want to solve it for now is that I apply a mark extension (like highlight) to the selection and I make sure to remove it when I'm done with these other steps. But I'm not very happy with this approach because it has a lot of pitfalls, like preserving these marks to the db, not removing them correctly, etc.

**Comment by kobrynovych** [comment]
@demiro  @abecirovic3 Hi, Is there a solution to this problem?

**Reply by abecirovic3** [reply]
I would recommend that you do some research on prosemirror decorations. I implemented a search highlight for my editor using decorations.

**Comment by kobrynovych** [comment]
found this code and updated it a bit and it seems to work fine

```

import { Extension } from '@tiptap/core'
import { Plugin, PluginKey } from '@tiptap/pm/state'
import { Decoration, DecorationSet } from '@tiptap/pm/view'

const DECO_NAME = 'onBlurHighlight'
const ACTION_TYPES = {
    BLUR: 'blur',
    FOCUS: 'focus',
}
export const OnBlurHighlight = Extension.create({
    name: DECO_NAME,

    addProseMirrorPlugins() {
        return [
            new Plugin({
                key: new PluginKey(DECO_NAME),

                state: {
                    init(_config) {
                        return DecorationSet.empty
                    },

                    apply: (transaction, oldState) => {
                        const { selection, doc } = transaction
                        const decoTransform = transaction.getMeta(DECO_NAME)
                        const hasSelection = selection && selection.from !== selection.to

                        if (!hasSelection || decoTransform?.action === ACTION_TYPES.FOCUS) {
                            return DecorationSet.empty
                        }

                        if (hasSelection && decoTransform?.action === ACTION_TYPES.BLUR) {
                            const decoration = Decoration.inline(selection.from, selection.to, {
                                class: 'blur-highlight',
                            })

                            return DecorationSet.create(doc, [decoration])
                        }

                        return oldState
                    },
                },
            
    props: {
                    decorations(state) {
                        return this.getState(state)
                    },
                    handleDOMEvents: {
                        blur: view => {
                            const { tr } = view.state

                            const transaction = tr.setMeta(DECO_NAME, {
                                from: tr.selection.from,
                                to: tr.selection.to,
                                action: ACTION_TYPES.BLUR,
                            })

                            view.dispatch(transaction)
                        },

                        focus: view => {
                            const { tr } = view.state

                            const transaction = tr.setMeta(DECO_NAME, {
                                from: tr.selection.from,
                                to: tr.selection.to,
                                action: ACTION_TYPES.FOCUS,
                            })

                            view.dispatch(transaction)
                        },
                    },
                },
            }),
        ]
    },
})

```

**Reply by abecirovic3** [reply]
Great, thanks for sharing

**Reply by bilalafzal01** [reply]
Thanks for sharing 💯

**Comment by lenguyenhoangkhang2** [comment]
I had the same issue and found a way to fix it! The key is to only show the decoration when the editor loses focus, and let the browser's selection color take over when focused.

Here's what worked for me:

```typescript
const SelectionDecoration = Extension.create({
  name: 'selectionDecoration',

  addOptions() {
    return {
      className: 'selection',
    }
  },

  addProseMirrorPlugins() {
    return [
      new Plugin({
        key: new PluginKey('selection'),
        props: {
          decorations: (state) => {
            const { selection } = state
            const { focused } = this.editor

            // Skip decoration if editor is focused or no selection
            if (focused || selection.empty) {
              return null
            }

            return DecorationSet.create(state.doc, [
              Decoration.inline(selection.from, selection.to, {
                class: this.options.className,
              }),
            ])
          },
        },
      }),
    ]
  },
})

**Comment by jackirvine97** [comment]
This is a very common use case in canvas style llms - suspect it will keep cropping up. Would be good to find a way to make this in built

**Reply by eviltik** [reply]
absolutely

**Comment by eviltik** [comment]
To keep system colors you can use these css values:

```css
background: Highlight;
color: HighlightText;
```

**Comment by barbaldo** [comment]
I'm also upping this with the rise of llms and such it is definitely a necessary core feature

**Comment by cactusanil** [comment]
any update over here

**Comment by arnaugomez** [comment]
First, have you checked the [Selection extension](https://tiptap.dev/docs/editor/extensions/functionality/selection)? It allows you to keep displaying the selection when you click outside the editor.

Another option is to store the range of the selection and recover it later.

You can get the range of the selection like this:

```ts
const selectionRange = editor.state.selection
```

A Range has two properties, `from` and `to`. These are the positions where the selection starts and ends.

Then, you can do many things with the range, like set the selection to that range and insert content there:

```ts
editor.chain().setTextSelection(selectionRange).insertContent('Example Text').run()
```

When the document is edited, the range changes. To keep the range updated as the document changes, you must listen to the `update` event ([see events guide](https://tiptap.dev/docs/editor/api/events)) and update the `from` and `to` positions using the [Tiptap position utility](https://tiptap.dev/docs/editor/api/utilities/position). 

If you want to display a special UI in the position of the `selectionRange`, [ProseMirror decorations](https://prosemirror.net/docs/ref/#view.Decorations) are the way to go about it. We're also working on an API for Tiptap decorations, we expect to release it soon. Stay tuned!
